L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc

```diff
@@ -0,0 +1,230 @@
+  void PhiFromXY(pxy_t px, pxy_t py, phi_t& phi, bool debug = false) const;
+
+  void CalcMetHLS(std::vector<float> pt, std::vector<float> phi, reco::Candidate::PolarLorentzVector& metVector) const;
+};
+
+L1MetPfProducer::L1MetPfProducer(const edm::ParameterSet& cfg)
+    : _l1PFToken(consumes<std::vector<l1t::PFCandidate>>(cfg.getParameter<edm::InputTag>("L1PFObjects"))),
+      maxCands_(cfg.getParameter<int>("maxCands")) {
+  produces<std::vector<l1t::EtSum>>();
+}
+
+void L1MetPfProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  edm::Handle<l1t::PFCandidateCollection> l1PFCandidates;
+  iEvent.getByToken(_l1PFToken, l1PFCandidates);
+
+  std::vector<float> pt;
+  std::vector<float> phi;
+
+  for (int i = 0; i < int(l1PFCandidates->size()) && (i < maxCands_ || maxCands_ < 0); i++) {
+    const auto& l1PFCand = l1PFCandidates->at(i);
+    pt.push_back(l1PFCand.pt());
+    phi.push_back(l1PFCand.phi());
+  }
+
+  reco::Candidate::PolarLorentzVector metVector;
+
+  CalcMetHLS(pt, phi, metVector);
+
+  l1t::EtSum theMET(metVector, l1t::EtSum::EtSumType::kTotalHt, 0, 0, 0, 0);
+
+  std::unique_ptr<std::vector<l1t::EtSum>> metCollection(new std::vector<l1t::EtSum>(0));
+  metCollection->push_back(theMET);
+  iEvent.put(std::move(metCollection));
+}
+
+void L1MetPfProducer::CalcMetHLS(std::vector<float> pt,
+                                 std::vector<float> phi,
+                                 reco::Candidate::PolarLorentzVector& metVector) const {
+  pxy_t hw_px = 0;
+  pxy_t hw_py = 0;
+  pxy_t hw_sumx = 0;
+  pxy_t hw_sumy = 0;
+
+  for (uint i = 0; i < pt.size(); i++) {
+    pt_t hw_pt = min(pt[i], maxPt_);
+    phi_t hw_phi = float(TVector2::Phi_mpi_pi(phi[i]) / phiLSB_);
+
+    Project(hw_pt, hw_phi, hw_px, true);
+    Project(hw_pt, hw_phi, hw_py, false);
+
+    hw_sumx = hw_sumx - hw_px;
+    hw_sumy = hw_sumy - hw_py;
+  }
+
+  pt2_t hw_met = pt2_t(hw_sumx) * pt2_t(hw_sumx) + pt2_t(hw_sumy) * pt2_t(hw_sumy);
+  hw_met = sqrt(int(hw_met));  // stand-in for HLS::sqrt
+
+  phi_t hw_met_phi = 0;
+  PhiFromXY(hw_sumx, hw_sumy, hw_met_phi);
+
+  metVector.SetPt(hw_met.to_double());
+  metVector.SetPhi(hw_met_phi.to_double() * phiLSB_);
+  metVector.SetEta(0);
+}
+
+void L1MetPfProducer::Project(pt_t pt, phi_t phi, pxy_t& pxy, bool isX, bool debug) const {
+  /*
+      Convert pt and phi to px (py)
+      1) Map phi to the first quadrant to reduce LUT size
+      2) Lookup sin(phiQ1), where the result is in [0,maxPt]
+      which is used to encode [0,1].
+      3) Multiply pt by sin(phiQ1) to get px. Result will be px*maxPt, but
+      wrapping multiplication is 'mod maxPt' so the correct value is returned.
+      4) Check px=-|px|.
+    */
+
+  // set phi to first quadrant
+  phi_t phiQ1 = (phi > 0) ? phi : phi_t(-phi);  // Q1/Q4
+  if (phiQ1 >= hwPiOverTwo_)
+    phiQ1 = hwPi_ - phiQ1;
+
+  if (phiQ1 > hwPiOverTwo_) {
+    std::cout << "unexpected phi (high)" << std::endl;
```

Comment:
Why do you use `std::cout` here? Use MessageLogger.

Not sure I understand the logic. You check `phiQ1` first, then update the value on the above `if`. Then check and update again here?